### ISSUE TITLE
Server doesn't read certificate?..

### ISSUE THREAD

**Issue by veox** [libbitcoin/libbitcoin-server#68]
`bs`: bb7afdc3794032287d22fbbc2e4647071f9d3ce1
`bx`: 7185278805fefdb1f7eacaf705ba8e065017012f

Built both today, changed `BITCOIN_OPTIONS` and `BITCOIN_BLOCKCHAIN_OPTIONS` in `install.sh` to have `--enable-testnet` for both.

Generated certificates with

``` sh
bx cert-new server.private
bx cert-new client.private
bx cert-public server.private server.public
bx cert-public client.private client.public
```

Edited configs:

``` ini
bs.cfg
---
[identity]
# The path to the ZPL-encoded server private certificate file.
cert_file = /home/veox/usr/etc/libbitcoin/server.private
# The directory for ZPL-encoded client public certificate files, allows anonymous clients if not set.
client_certs_path = /home/veox/usr/etc/libbitcoin
```

``` ini
bx.cfg
---
[testnet]
# The URL of the Libbitcoin/Obelisk testnet server.
url = tcp://localhost:9091
# The Z85-encoded public key of the server certificate.
server_cert_key = N3EY58Dy:K-PMiiZObPa[53EO[/vq@wkBNiBkG}R
# The path to the ZPL-encoded client private certificate file.
cert_file = /home/veox/usr/etc/libbitcoin/client.private
```

Running `bs`, and then `bx fetch-height` results in:

`LIBZMQ I: security failure, self=CURVE peer=NULL`

On server's side, several (2-4) lines of this are shown:

`LIBZMQ I: security failure, self=NULL peer=CURVE`

So it seems the server doesn't load its certificate.

In addition to that, can't check what server thinks it has loaded - `bs --settings` results in:

```
Configuration Settings:

[endpoints]
# The block publishing service endpoint, defaults to 'tcp://*:9093'.
block_publish = <VALUE>
# The heartbeat endpoint, defaults to 'tcp://*:9092'.
heartbeat = <VALUE>
# The query service endpoint, defaults to 'tcp://*:9091'.
service = <VALUE>
# The transaction publishing service endpoint, defaults to 'tcp://*:9094'.
tx_publish = <VALUE>

[general]
# The blockchain directory, defaults to 'blockchain'.
blockchain_path = <VALUE>
# The minimum height of the history database, defaults to 0.
history_height = <VALUE>
# The peer cache file path, defaults to 'hosts'.
hosts_file = <VALUE>

...
```

**Comment by veox** [comment]
Adding a breakpoint at https://github.com/libbitcoin/libbitcoin-server/blob/master/src/worker.cpp#L85 shows that `bs` steps into the "if" clause.

**Comment by veox** [comment]
Further inversigation shows default config file name has remained "server.cfg".

**Comment by evoskuil** [comment]
So does this mean certs are working as expected, in both directions?

**Comment by veox** [comment]
I've tried the following:
* With both configs populated with correct values, queries are replied to properly (no errors).
* If `bs` requires authenticated clients (by having a set `client_certs_path`) and `bx` doesn't provide one, the query fails (as expected). `bs` shows several `CURVE I: cannot open client INITIATE vouch` errors, until `bx` times out.
** However, if `bx` timeout is set extremely high, `bs` log gets spammed. It might be prudent for `bs` to close the connection instead.
** It might be nice to have an infinite timeout for `bx` if `wait` is set to `0`. ATM `bx` does precisely what is said - times out immediately, without sending a query.
* If `bs` has a cert set, but `bx`doesn't have a `server_cert_key` set, then the query times out.
** This is unexpected, however understandable, since `bx` doesn't have a public key to use. Perhaps it should show a warning.
* If `bs` doesn't have a `client_certs_path` set, and `bx`has a cert, then the query succeeds (as expected).
* If `bs` doesn't have a `cert_file` set, but `bx` has a server key, then query fails with `LIBZMQ I: security failure, self=CURVE peer=NULL` (as expected).

**Comment by evoskuil** [comment]
Thanks for the thorough review.

We did look at having a max timeout if timeout was set to zero. That was when the default was zero, but with the default at 2000 ms it doesn't seem like a  problem. The user would have to explicitly make the timeout 0 ms to see the behavior, in which case it is correct. In fact if the client API worked that way, implementers would have to handle it as a special case. A rounding difference could change a small timeout into max timeout. In any case we probably wouldn't use infinity, but the max value is quite a long period of time (49 days).

I agree with the closing of the connection at the server and the client warning if the server has no cert when expected.

**Comment by evoskuil** [comment]
I've gone through an extensive evaluation of client-server behavior under all configurations and verified correctness in master (version3).

**Comment by evoskuil** [comment]
The timeout issues were a result of incorrect wait/retry implementation in `libbitcoin-client`.
